### DISCUSSION TITLE
API rules for one-to-many relation

### DISCUSSION THREAD

**Discussion by elliotsoomro** [pocketbase/pocketbase#1272] [Q&A]
I have two collections: questions and alternatives. Questions have a relation field "alternatives" and a "user" relation field which is used to define API rules (@request.auth.id = user.id).

Now I want to define API rules for the alternatives collection, but how do I check if the question that the alternative is associated with is owned by the user? With this table structure I realize that an alternative could be associated to multiple questions, but that is not necessary.

I have tried using the "@collection", but it does not seem to solve this problem (@collection.questions.alternatives.id = id && @collection.questions.user.id = @request.auth.id).

One option could be keeping a "question" relation field on alternatives. Is that the best solution? It would require first creating the alternatives, then creating the question with associated alternatives, and then updating the "question" relation field on the alternatives.

To efficently being able to query both question "text" field and alternative "text" field at the same time it is necessary to have the "alternatives" relation field on questions.

**Comment by ganigeorgiev** [comment]
> I have tried using the "@collection", but it does not seem to solve this problem (@collection.questions.alternatives.id = id && @collection.questions.user.id = @request.auth.id).

The above rule should work, but it may not return what you expect, since, if I understand correctly, the `alternatives` is a multi `relation` field and the rule will result to true if there is at least one alternative matching the condition (internally we are doing `LEFT JOIN`). I'm currently working on multi-match expressions but this would take a while.

---

> One option could be keeping a "question" relation field on alternatives. Is that the best solution? 

This is the more "traditional" SQL approach, aka. the relation flow would be something like: `alternatives->questions->user`.
Whether it is the best solution - I'm not sure, it depends on your use case. This approach has one other benefit and that is that you can remove/add alternatives without providing the other existing rel ids since it will be a single `relation` field (field modifiers for multi-relation is in the roadmap).

> It would require first creating the alternatives, then creating the question with associated alternatives, and then updating the "question" relation field on the alternatives.

I'm not sure that I understand the described flow. The illustrated example seems to be for a multi-relation field. If you move the relation to the `alternatives` collection and make it single (aka. "Max Select = 1"), the flow would be:

```js
const question = await pb.collection("questions").create({
    text: "example question..."
});

const alternative1 = await pb.collection("alternatives").create({
    text: "example alternative...",
    question: question.id,
})

const alternative2 = await pb.collection("alternatives").create({
    text: "example alternative...",
    question: question.id,
})

...
```

---

> To efficently being able to query both question "text" field and alternative "text" field at the same time it is necessary to have the "alternatives" relation field on questions.

If you are not concerned about the network traffic, making many smaller requests may be more efficient than one big complex query (especially when using SQLite).
You can query both alternatives and questions only through the main collection (in this case alternatives; "indirect" relations filtering is not supported at the moment):

```
?filter=(text ~ 'alternative text' || question.text ~ 'question text')
```
